vault/resource_kmip_secret_role.go

```diff
@@ -0,0 +1,318 @@
+	client := meta.(*api.Client)
+	rolePath := d.Id()
+	if rolePath == "" {
+		return fmt.Errorf("expected a path as ID, got empty string")
+	}
+
+	log.Printf("[DEBUG] Reading KMIP role at %q", rolePath)
+	resp, err := client.Logical().Read(rolePath)
+	if err != nil {
+		return fmt.Errorf("error reading KMIP role at %s, err=%w", rolePath, err)
+	}
+	if resp == nil {
+		log.Printf("[WARN] KMIP role not found, removing from state")
+		d.SetId("")
+
+		return fmt.Errorf("expected role at %s, no role found", rolePath)
+	}
+
+	for _, k := range []string{fieldTLSClientKeyType, fieldTLSClientKeyBits, fieldTLSClientTTL} {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return err
+		}
+	}
+
+	for _, k := range kmipRoleAPIBooleanFields {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return fmt.Errorf("error setting state key %q on KMIP role, err=%w", k, err)
+		}
+	}
+
+	return nil
+}
+
+func kmipSecretRoleUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	rolePath := d.Id()
+
+	if d.HasChange("path") {
+		newRolePath := getRolePath(d)
+
+		log.Printf("[DEBUG] Confirming KMIP role exists at %s", newRolePath)
+		resp, err := client.Logical().Read(newRolePath)
+		if err != nil {
+			return fmt.Errorf("error reading role at path %s, err=%w", newRolePath, err)
+		}
+
+		if resp == nil {
+			return fmt.Errorf("error remounting KMIP role to new backend path %s, err=%w", newRolePath, err)
+		}
+		d.SetId(newRolePath)
+		rolePath = newRolePath
+	}
+
+	data := kmipSecretRoleRequestData(d)
+	log.Printf("[DEBUG] Updating %q", rolePath)
+
+	if _, err := client.Logical().Write(rolePath, data); err != nil {
+		return fmt.Errorf("error updating KMIP role %q,  err=%w", rolePath, err)
+	}
+	log.Printf("[DEBUG] Updated %q", rolePath)
+
+	return kmipSecretRoleRead(d, meta)
+}
+
+func kmipSecretRoleDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	rolePath := d.Id()
+
+	log.Printf("[DEBUG] Deleting KMIP role %s", rolePath)
+	_, err := client.Logical().Delete(rolePath)
+	if err != nil {
+		return fmt.Errorf("error deleting role %s", rolePath)
+	}
+	log.Printf("[DEBUG] Deleted KMIP role %q", rolePath)
+
+	return nil
+}
+
+func kmipSecretRoleRequestData(d *schema.ResourceData) map[string]interface{} {
+	nonBooleanfields := []string{fieldTLSClientKeyType, fieldTLSClientKeyBits, fieldTLSClientTTL}
+
+	data := make(map[string]interface{})
+	for _, k := range nonBooleanfields {
+		if d.IsNewResource() {
+			if v, ok := d.GetOk(k); ok {
+				data[k] = v
+			}
+		} else if d.HasChange(k) {
+			data[k] = d.Get(k)
+		}
+	}
+
+	// Boolean fields must not be evaluated
+	// otherwise all result in true
+	for _, k := range kmipRoleAPIBooleanFields {
+		if d.IsNewResource() {
+			if v, ok := d.GetOkExists(k); ok {
+				data[k] = v.(bool)
+			}
+		} else if d.HasChange(k) {
+			data[k] = d.Get(k)
+		}
+	}
+
+	return data
+}
+
+func getRolePath(d *schema.ResourceData) string {
```

Comment:
Since the `vault` package is quite massive, we should probably rename this to to something like: `getKMIPRolePath()`
```suggestion
func getKMIPRolePath(d *schema.ResourceData) string {
```

Reply:
Fixed in 59bd3f5